File: visitor_pattern/C/src/merge_sort.cpp
#include <inc/merge_sort.h>
#include <algorithm>

template<typename T, typename C>
void merge_sort<T, C>::divide(size_t s, size_t e)
{
   if ( s + 1 >= e)
       return; 
   size_t m = (s+e)/2;
   divide(s, m);
   divide(m, e);
   merge(s, e);
}
template<typename T, typename C>
void merge_sort<T,C>::merge(size_t s, size_t e)
{
    C c;
    size_t m = (s+e)/2, i = s, j = m; 
    size_t k = 0;
    while(i < m && j < e)
        h[k++] = c(a[j], a[i])? a[i++]: a[j++];
    while(i < m)
        h[k++] = a[i++];  
    while(j < e)
        h[k++] = a[j++];
    std::copy(h, h + e - s, a + s);
}
template<typename T, typename C>
void merge_sort<T,C>::do_work(void)
{
   divide(0, sz);
}

File: visitor_pattern/java/MergeSort.java
//package com.charlesq.java.visitorpattern;
import java.lang.Comparable;
import java.lang.Class;
import java.lang.reflect.Array;
import java.util.Arrays;
/* a concrete element in visitor pattern */
public class MergeSort<T extends Comparable<T>> implements Visitable<T>
{
    private transient T [] b;
    private transient T[] a;
    @Override
    public void accept (Visitor<T> v)
    {
       v.visit(this); 
    }
    public void init(T [] a)
    {
        this.a = a;
    }
    public void run()
    {
        b  = ( T[])Array.newInstance(a.getClass().getComponentType(), a.length);
        divide(0, a.length);
    }
    @Override
    public String toString()
    {
        return "Merge Sort";
    }
    private void divide(int s, int e)
    {
        /* termination condition in recursive divide */
        if (s +1 >= e)
            return;
         divide(s, (s+e)>>>1);
         divide ((s+e)>>>1, e);
         merge(s, e);
    }
    private void merge(int s, int e)
    {
        int m = (s + e) >>> 1, i = s, j = m, k = 0;
        while(i < m && j < e)
            b[k++] = (a[i].compareTo(a[j]) <= 0)? a[i++]: a[j++];
        while(i < m)
            b[k++] = a[i++];
        while(j < e)
            b[k++] = a[j++];
        while (k-- != 0)
            a[--e] = b[k];
    }
}

 
